File: services/api-py/scripts/check_boundaries.py
"""The import-boundary rule for this service, checked over every source file instead of left to review.

Each layer lists what it MAY import. An allowlist, because a denylist only forbids the mistakes its
author already thought of:

    api_py/domain       its own package, and only standard-library modules that are themselves pure
    api_py/application  the domain, its own package, and the same pure modules
    api_py/adapters     the domain, the application, its own package, any module, any package
    api_py/main.py and api_py/config.py are the composition root and may import anything
    any other module under api_py/ belongs to no layer, and fails

Imports are read with ``ast``, the same parser Python itself uses, so a name inside a string or a
comment cannot raise a false alarm and no import can hide from it.

    uv run python scripts/check_boundaries.py

Exit 0 clean, 1 violations.
"""

from __future__ import annotations

import ast
import sys
from dataclasses import dataclass
from pathlib import Path

PACKAGE = "api_py"

# Standard-library modules with no I/O, no clock and no randomness. Everything else — os, time,
# random, json, threading, pathlib — belongs to an adapter or the composition root.
PURE_STDLIB = frozenset({"__future__", "abc", "collections", "collections.abc", "dataclasses", "enum", "re", "typing"})

COMPOSITION_ROOT = frozenset({"main.py", "config.py"})


@dataclass(frozen=True, slots=True)
class Layer:
    package: str
    may_import: frozenset[str]
    anything: bool = False


LAYERS = (
    Layer("domain", frozenset({"domain"})),
    Layer("application", frozenset({"domain", "application"})),
    Layer("adapters", frozenset({"domain", "application", "adapters"}), anything=True),
)


def imported_modules(source: str, relative: str) -> list[str]:
    """Every module this source imports, as an absolute dotted name.

    A relative import is resolved against the importing file first. `from .other import x` stays in
    the layer, but `from ..adapters import x` leaves it, and only resolution can tell them apart.
    """
    tree = ast.parse(source)
    here = f"{PACKAGE}.{relative.rsplit('/', 1)[0]}" if "/" in relative else PACKAGE
    names: list[str] = []
    for node in ast.walk(tree):
        if isinstance(node, ast.Import):
            names.extend(alias.name for alias in node.names)
        elif isinstance(node, ast.ImportFrom):
            if node.level == 0:
                names.append(node.module or "")
                continue
            parts = here.split(".")
            base = parts[: len(parts) - (node.level - 1)]
            names.append(".".join([*base, *([node.module] if node.module else [])]) or PACKAGE)
    return [name for name in names if name != ""]


def check_file(relative: str, source: str) -> list[str]:
    """Violations in one file. ``relative`` is POSIX and rooted at the package, e.g. domain/task.py."""
    head = relative.split("/")[0]
    layer = next((one for one in LAYERS if one.package == head), None)
    if layer is None:
        if relative in COMPOSITION_ROOT:
            return []
        return [
            f"{relative} belongs to no layer. Move it into "
            f"{', '.join(one.package for one in LAYERS)}, or add a layer with its own allowlist."
        ]

    problems: list[str] = []
    for module in imported_modules(source, relative):
        if module == PACKAGE or module.startswith(f"{PACKAGE}."):
            target = module[len(PACKAGE) + 1 :].split(".")[0]
            if target not in layer.may_import:
                problems.append(
                    f"{relative} imports {module}; {layer.package} may import only "
                    f"{', '.join(sorted(layer.may_import))}."
                )
        elif not layer.anything and module.split(".")[0] not in PURE_STDLIB and module not in PURE_STDLIB:
            problems.append(
                f"{relative} imports {module}; {layer.package} performs no I/O and may import only "
                f"pure modules ({', '.join(sorted(PURE_STDLIB))}). Put the integration in an adapter."
            )
    return problems


def check_tree(root: Path) -> tuple[int, list[str]]:
    package_root = root / "src" / PACKAGE
    files = sorted(path for path in package_root.rglob("*.py") if path.name != "__init__.py")
    problems: list[str] = []
    for path in files:
        relative = path.relative_to(package_root).as_posix()
        problems.extend(check_file(relative, path.read_text(encoding="utf-8")))
    return len(files), problems


def main() -> int:
    files, problems = check_tree(Path(__file__).resolve().parent.parent)
    if problems:
        print(f"check-boundaries: {len(problems)} violation(s)\n", file=sys.stderr)
        for problem in problems:
            print(f"  {problem}", file=sys.stderr)
        return 1
    print(f"check-boundaries: OK. {files} files, 0 violations.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: services/mcp-server/scripts/check-boundaries.mjs
/**
 * The import-boundary rule for this service, checked over every source file instead of left to review.
 *
 * Each layer lists what it MAY import. An allowlist, because a denylist only forbids the mistakes its
 * author already thought of:
 *
 *   src/domain       its own files only; no packages and no node: built-ins (no I/O, clock or randomness)
 *   src/application  the domain and its own files
 *   src/adapters     the domain, the application, its own files, node: built-ins and packages
 *   src/main.ts and src/config.ts are the composition root and may import anything
 *   any other file under src/ belongs to no layer, and fails
 *
 * Adapted from NexusPrompt's check-boundaries.mjs. Specifiers are read with a regular expression, not
 * a parser: static import/export-from, side-effect imports and dynamic import() with a string literal.
 * An import-shaped string inside a comment can raise a false alarm; nothing makes it miss a real import.
 *
 *   node scripts/check-boundaries.mjs
 *
 * Exit 0 clean · 1 violations.
 */
import { readdirSync, readFileSync, statSync } from "node:fs";
import { dirname, join, posix, relative, resolve } from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";

export const LAYERS = [
  { dir: "src/domain", mayImport: ["src/domain"], builtins: false, packages: false },
  { dir: "src/application", mayImport: ["src/domain", "src/application"], builtins: false, packages: false },
  { dir: "src/adapters", mayImport: ["src/domain", "src/application", "src/adapters"], builtins: true, packages: true },
];

const IMPORT_RE =
  /(?:^|[\n;])\s*(?:import|export)\b[^'"`;]*?\bfrom\s*["']([^"']+)["']|(?:^|[\n;])\s*import\s*["']([^"']+)["']|\bimport\s*\(\s*["']([^"']+)["']\s*\)/g;

/** The composition root may import anything. Every other file under src/ must belong to a layer. */
export const COMPOSITION_ROOT = ["src/main.ts", "src/config.ts"];

export const specifiers = (source) => [...source.matchAll(IMPORT_RE)].map((m) => m[1] ?? m[2] ?? m[3]);

/** Violations in one file. `file` is a POSIX path relative to the service root, such as src/domain/task.ts. */
export function checkFile(file, source) {
  const layer = LAYERS.find((l) => file.startsWith(`${l.dir}/`));
  if (layer === undefined) {
    // A file in no layer is a file no rule covers; a new directory must not escape by being new.
    return COMPOSITION_ROOT.includes(file)
      ? []
      : [`${file} belongs to no layer. Move it into ${LAYERS.map((l) => l.dir).join(", ")}, or add a layer with its own allowlist.`];
  }
  const problems = [];
  for (const spec of specifiers(source)) {
    if (spec.startsWith(".")) {
      const target = posix.normalize(posix.join(posix.dirname(file), spec));
      if (!layer.mayImport.some((dir) => target.startsWith(`${dir}/`))) {
        problems.push(`${file} imports "${spec}" (${target}); ${layer.dir} may import only from ${layer.mayImport.join(", ")}.`);
      }
    } else if (spec.startsWith("node:")) {
      if (!layer.builtins) problems.push(`${file} imports "${spec}"; ${layer.dir} performs no I/O and may not use Node built-ins.`);
    } else if (!layer.packages) {
      problems.push(`${file} imports package "${spec}"; ${layer.dir} may not depend on packages. Put the integration in an adapter.`);
    }
  }
  return problems;
}

function sourceFiles(root, dir) {
  const out = [];
  for (const entry of readdirSync(join(root, dir))) {
    const path = posix.join(dir, entry);
    if (statSync(join(root, path)).isDirectory()) out.push(...sourceFiles(root, path));
    else if (/\.(ts|tsx|mts|js|mjs)$/.test(entry)) out.push(path);
  }
  return out;
}

export function checkTree(root) {
  const files = sourceFiles(root, "src");
  return { files: files.length, problems: files.flatMap((file) => checkFile(file, readFileSync(join(root, file), "utf8"))) };
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  const root = resolve(dirname(fileURLToPath(import.meta.url)), "..");
  const { files, problems } = checkTree(root);
  if (problems.length > 0) {
    console.error(`check-boundaries: ${problems.length} violation(s)\n\n  ${problems.join("\n  ")}`);
    process.exitCode = 1;
  } else {
    console.log(`check-boundaries: OK. ${files} files in ${relative(process.cwd(), root) || "."}, 0 violations.`);
  }
}
